kusto/ingest/ingest.go

```diff
@@ -1,22 +1,24 @@
 package ingest
```

Comment:
I had this conversation yesterday with Yochai in that this managed thing looks like something that should just be rolled into the existing structure or replace it.  This managed thing also seems to be the frontend trying to fix a backend problem.   I guess it is what it is.....

I'm not as interested in how this looks in the other APIs as much as making sure the user can use the feature.  And I'm super interested in a clean API with the least amount of junk.  We should avoid unnecessary API breakage, but we are a 0.X release, so we can do so if the benefit is high enough.

@AsafMah : I'd like to think about how to roll as many of these methods together as possible.  Unless we MUST keep them separate because of functional reasons, I'd like to combine these and not multiply them.  We can use options to specify specific settings (like Managed()). 

My ask is:
@AsafMah : forgetting the code implementation, can you propose an API with the least moving parts that supports our current cases + managed?

Reply:
My big problem with making this an option is discoverability, most people won't know that managed ingestion is a thing and probably won't use it.

Other SDKs have three separate client types for different types of ingestion - Queued, Streaming and ManagedStreaming
Since we only have one client in go that supports both streaming and queued, I thought that adding managed functions is a good middleground.

The options as I see them are -
1. Split it to a different client struct - we don't want that
2. Have "managed" be an additional parameter to the functions we have - go doesn't support default parameters or overloading, so this breaks the interface
3. Add managed as a FileOption - Discoverability + it's not really a "file option"
4. Add it as an option and make it the default - we probably don't want to change the default behaviour
4. Add new Managed functions - I agree it's not ideal, but I think it's best considering the circumstances

Reply:
@element-of-surprise , since you already plan on breaking and remaking the SDK, lets have two APIs calling one hidden impl.
When you've rearchitected the SDK, feel free to drop the old streaming API.
WDYT?

Reply:
@AsafMah 

#3 seems more ideal to me, and I don't agree with the premises in that discoverability is a problem.  A Go user is either going to look at the options if they are interested or search for "Managed" in the docs if they know its there.

#4 you can do default values in Go, you just don't do it with "attribute=Default" in the definition of some method.

FileOption was chosen as the name when we just had Stream and FromFile().  It was notating that it just changed files, not that it matters all that much. The user just looks at the type and then looks for what implements the options (in Go).  I honestly don't even see the type names for options.

My mind is starting to coalesce around a single main method for doing everything, with options.  That method would be:

```go
 FromReader(ctx context.Context, reader io.Reader, options ...ReaderOption) (*Result, error)
```

FromReader, given the right arguments, can do everything.  This could handle input from a stream, memory, or a file.  Given the right options.  To me, Managed() just becomes an option at that point.  

We could then add our individual FromFile(), FromStream() directly on top of that or we could fork out a "high level" package that had these.  They would have their own options that would put in ReaderOption(s).  Some of those options might be "hidden"  for FromReader buy exposed by our high levels.

@yogilad 

So it depends on how disruptive you want to be.  What you are talking about doing is disrupting in a few different ways:

If we make changes now:
FromReader() (or maybe just calling it From()) having some minor changes.
FromFile() having minor changes, possibly a location move
FromStream() having minor changes, possibly a location move

If we do it later:
Removal of Managed* + those items above.  
User must figure out that Managed* became an option and those methods are gone.

@yogilad and @AsafMah 

We probably need to sit down and discuss this.  We are coming at this from two different directions with the same goal.  I'm super hard on APIs, because changes are usually either disruptive or add baggage.  

If managed is a thing our clients need, then I want it in the API.

But I don't want to start exploding the API surface to get there.  I heard something once that resinates with me: " if you know something well, you don't add features, you take them away.  "  I'm paraphrasing, but I think it goes with "Less is More" from Pike:  https://commandcenter.blogspot.com/2012/06/less-is-exponentially-more.html